documentation/architecture/ADR1:dependency-injection.md

```diff
@@ -0,0 +1,76 @@
+## Context
+##### What is DI?
```

Comment:
I think the formatting of this document is a little confusing if you're coming in with no context:
- it doesn't separate the "Context" from the "Decision" (it's missing the "decision" section entirely)
- It doesn't clearly identify the problems *our team* is having
- It doesn't clearly identify why the chosen solution better solves our problems than another solution
- I think it leaves out some potential future problems with ServiceLocator (e.g. when SL grows large, it's hard to maintain, no compile-time checks for dependency graph, etc.) and why we don't care about these problems

In general, I think the formatting doesn't make it clear what options there are and the trade-offs between them (e.g. DI vs. ServiceLocator talks trade-offs but then is following by "Why not DI framework?" and "Why not Singletons?" where DI leaks into two sections, the two "Why not" sections aren't comparing similar things, and Singletons wasn't mentioned as an option sooner).

I think you should start with the specific problems that *our team* is having which is why we implemented this solution and be clear to enumerate what our options are (e.g. in lists), e.g.:

> ## Context
> Our code is currently using singletons for blah blah. We're having problems testing our code because they make hard-coded to make network calls and are tightly coupled to their dependencies.
>
> The industry-standard way of solving this problem is the Dependency Injection (wikipedia-link) pattern. The DI pattern solves this issue by passing the dependencies into a class, wherein they can be mocked. There are no other standard solutions to this problem. (sample of with and without DI).
> 
> Currently, there are X standard ways to do the dependency injection pattern:
> - ServiceLocator
> - Dependency Injection framework
> - Application holds singletons
>
> (Briefly describe differences, trade-offs)
>
> ## Decision
> Given the problems we were having, we decided on X because these trade-offs are worth it and we don't care about trade-offs Y, and Z.